src/jobserver-posix.cc

```diff
@@ -0,0 +1,170 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "jobserver.h"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <cassert>
+#include <cstring>
+
+#include "util.h"
+
+Jobserver::Jobserver() {
+  assert(!Enabled());
+
+  // Return early if no makeflags are passed in the environment.
+  char* makeflags = std::getenv("MAKEFLAGS");
+  if (makeflags == nullptr || strlen(makeflags) == 0)
+    return;
+
+  std::string::size_type flag_char_ = 0;
+  std::string flag_;
+  std::vector<std::string> flags_;
+
+  // Tokenize string to characters in flag_, then words in flags_.
+  while (flag_char_ < strlen(makeflags)) {
+    while (flag_char_ < strlen(makeflags) &&
+           !isblank(static_cast<unsigned char>(makeflags[flag_char_]))) {
+      flag_.push_back(static_cast<unsigned char>(makeflags[flag_char_]));
+      flag_char_++;
+    }
+
+    if (!flag_.empty())
+      flags_.push_back(flag_);
+
+    flag_.clear();
+    flag_char_++;
+  }
+
+  // --jobserver-auth=<val>
+  for (size_t n = 0; n < flags_.size(); n++)
+    if (flags_[n].find(AUTH_KEY) == 0)
+      flag_ = flags_[n].substr(strlen(AUTH_KEY));
+
+  // --jobserver-fds=<val>
+  if (flag_.empty())
+    for (size_t n = 0; n < flags_.size(); n++)
+      if (flags_[n].find(FDS_KEY) == 0)
+        flag_ = flags_[n].substr(strlen(FDS_KEY));
+
+  // -j 1
+  if (flag_.empty())
+    for (size_t n = 0; n < flags_.size(); n++)
+      if (flags_[n].find("-j") == 0)
+        jobserver_closed_ = true;
+
+  jobserver_name_.assign(flag_);
+  const char* jobserver = jobserver_name_.c_str();
+
+  // Check for fifo type first.
+  if (jobserver_name_.find(FIFO_KEY) == 0)
+    jobserver_fifo_ = true;
+
+  // Warn if jobserver type is unknown (neither fifo nor pipe).
+  if (!jobserver_fifo_ && sscanf(jobserver, "%d,%d", &rfd_, &wfd_) != 2)
+    if (!jobserver_name_.empty())
+      Warning("invalid jobserver value: '%s'", jobserver);
+
+  // Open FDs to the pipe if needed.
+  if (jobserver_fifo_) {
+    rfd_ = open(jobserver + strlen(FIFO_KEY), O_RDONLY | O_NONBLOCK);
+    wfd_ = open(jobserver + strlen(FIFO_KEY), O_WRONLY);
+  }
+
+  // Exit on failure to open FDs, build non-parallel for invalid passed FDs.
+  if (rfd_ >= 0 && wfd_ >= 0)
+    Info("using jobserver: %s", jobserver);
+  else if (jobserver_fifo_ && (rfd_ == -1 || wfd_ == -1))
+    Fatal("failed to open jobserver: %s: %s", jobserver, strerror(errno));
+  else
+    jobserver_closed_ = true;
+
+  // Fallback to non-parallel if jobserver-capable parent has no jobserver.
+  if (jobserver_closed_)
+    rfd_ = wfd_ = 0;
```

Comment:
did you mean -1 here?

Reply:
setting the FDs to 0 when the pipe is closed is to ensure that Enabled() still returns true so that the default parallelism of `processors + 2` or whatever the user sets for `-j` is blocked. 

I was trying to avoid changing the definition of Enabled() to anything other than the FD numbers, but if you think it looks better the function can be defined as:

```
bool Jobserver::Enabled() const {
  return (rfd_ >= 0 && wfd_ >= 0) || jobserver_closed_;
}
```

then there would be no need to set the FDs here, although that might also look counterintuitive...

Reply:
Frankly, this is the kind of clever coding that ends up creating confusion or subtle bugs over time. At a minimum there should be a comment over any subtle code paths when they are really required (e.g. for performance reason).

In this specific case, I recommend you use_ an `enabled_` boolean flag instead, and set the descriptors to -1.

Reply:
it feels somewhat bad (harder to detect errors, or leading to potentially unusual output) when using numbers in FD variables which are strictly valid file descriptors. It is probably better to use an explicit negative magic constant instead.

Reply:
> did you mean -1 here?

this has been reworked, and comments added for `jobserver_closed_` in the header.